FEATURE: Add actor deletion

angusmcleod opened the pull request:
@pmusaraj In the course of working on this I found it difficult to keep things straight without cleaning up the `plugin.rb` a bit and applying zeitwerk loading, so those changes (planned for this round of work) are in this too. For the substantive changes in this PR, see these specs in particular:

- spec/lib/discourse_activity_pub/ap/activity/delete_spec.rb
- spec/lib/post_destroyer_spec.rb
- spec/models/category_spec.rb
- spec/models/tag_spec.rb
- spec/requests/admin/discourse_activity_pub/actor_controller_spec.rb

For context on the approach to deletion, see further:
- https://socialhub.activitypub.rocks/t/reconciling-activitypub-deletes-with-nodebb-deletion/4221/12?u=angus
- https://codeberg.org/fediverse/fep/src/branch/main/fep/4f05/fep-4f05.md

pmusaraj commented:
Sorry for the delay @angusmcleod, I'll review this shortly.

pmusaraj commented on assets/javascripts/discourse/controllers/admin-plugins-activity-pub-actor-show.js:
```diff
@@ -60,27 +63,41 @@ export default class AdminPluginsActivityPubActorShow extends Controller {
+
+  @action
+  deleteActor() {
+    this.dialog.yesNoConfirm({
```
Can we use a `deleteConfirm` dialog here? We can make `Are you sure you want to delete ACTOR_NAME?` a title, and have the action button use the `danger` color, it's more appropriate for a delete action. 

Also, in the warning message, these two lines are unclear: 

- Posts authored by apb@d1.musaraj.com will be deleted.

Does this refer to ActivityPub posts? I guess so but we should clarify, so that admins don't think that the actor's Discourse posts will get deleted. 

- Other "APB" posts will not be affected.

This is unclear too.

angusmcleod replied:
Good feedback, thanks. I've made various changes to the delete and destroy confirmation modals along the lines you've identified: https://github.com/angusmcleod/discourse-activity-pub/commit/8f2f194252fe643b05fb8c910cb0a65aa15e5d10

pmusaraj commented on lib/discourse_activity_pub/actor_handler.rb:
```diff
@@ -121,15 +121,28 @@ def self.ensure_required_attributes(actor)
     end
 
+    def self.delete_user(user, destroy: false)
+      if destroy
```
In what context do we need to destroy the user record here? Kinda worried this can unintentionally lead to destroyed and unrecoverable user records. Would much prefer to leave the decision here to core.

angusmcleod replied:
The `destroy` flag will be true when we:

1. receive a `Delete` of an `Actor`; and
2. are unable to resolve the `Actor`, i.e. it is remote and inaccessible.

This is the behaviour I worked out with Julian (NodeBB). See the spec here: https://codeberg.org/fediverse/fep/src/branch/main/fep/4f05/fep-4f05.md#http-404-or-410. I've added an additional `remote?` check into the callback where the `destroy` flag is set for clarity (`activity.object.cache["delete_object"]` will also only be true when it's a remote inaccessible actor, but additional clarity / checks here won't hurt).

In those circumstances, i.e. a user only associated with a deleted remote actor, I think a direct destroy of the associated `user` is warranted as that user is not really controlled by the local instance. 

But I think you have highlighted an important case, namely when a remote actor is associated with user that is in fact used by a real person, e.g. when a user associates their account with a remote actor using our authorisation features. We don't want to destroy the user in those circumstances if the remote actor is deleted. I've added in the following additional checks to the destroy flag:

1. Double checking the user has a remote actor with it.
2. User is staged, i.e. it has been automatically created by the AP plugin.

So, in summary, the AP plugin will only destroy users in these circumstances:
1. We receive a `Delete` of a remote `Actor`.
2. We are unable to resolve the remote `Actor`.
3. Our record of the deleted, remote, inaccessible, `Actor` has a staged `user` associated with it.

Commit: https://github.com/angusmcleod/discourse-activity-pub/commit/568644fcbf522b73d6f1731d461f75ca04e2574c

pmusaraj replied:
Great, thanks for the detailed explanation! 

> Our record of the deleted, remote, inaccessible, Actor has a staged user associated with it.

Is it at all possible that the same staged user could be in the instance for other reasons? Core creates staged users on various circumstances, for example when sending a PM to an email address. I can imagine this is rare, but I wonder if there is a scenario where we'd delete a valid staged user.

pmusaraj reviewed (commented):
A few early comments, will return to this tomorrow, quite hard to go over everything in one shot with PRs of this size.

pmusaraj reviewed (approved):
Overall this looks good to me. I will try to merge first thing tomorrow. 

The only remaining question is regarding destroying staged users. Would like to be very careful with that, it's an irreversible change.

angusmcleod commented:
I've added an additional spec for staged users without AP actors to ensure they won't be affected by this: https://github.com/angusmcleod/discourse-activity-pub/commit/2e2929060eb3b2c7c74a9265e8cb591397029522.

I've also added an additional check that the staged users we're destroying don't have any emails associated with them. If they have a remote AP actor (one of the other checks) then they won't have an email, but let's add in this check in case things change in the future.

I've also added a new user custom field identifying users auto-created by the AP plugin. This also means that any destruction of AP staged users will only be prospective, and won't affect any existing users.

So, in summary, the only scenario in which a staged user will be destroyed is if:
1. We receive a Delete of a remote Actor.
2. We are unable to resolve the remote Actor.
3. The remote Actor has a staged user associated with it.
4. The staged user does not have any emails associated with it.
5. The staged user has an `activity_pub_user` custom field.

Note that we're also setting the appropriate `context` when we perform a staged user destruction, so we'll know if the user was destroyed by the AP plugin, i.e. the context will be: "User destroyed by ActivityPub Delete Activity: %{username}".
